consensus/src/dag/round_state.rs

```diff
@@ -0,0 +1,179 @@
+    async fn check_for_new_round(
+        &mut self,
+        highest_strong_links_round: Round,
+        strong_links: Vec<NodeCertificate>,
+    );
+
+    fn reset(&mut self);
+}
+
+/// Move as fast as 2f+1
+pub struct OptimisticResponsive {
+    event_sender: tokio::sync::mpsc::Sender<Round>,
+}
+
+impl OptimisticResponsive {
+    pub fn new(event_sender: tokio::sync::mpsc::Sender<Round>) -> Self {
+        Self { event_sender }
+    }
+}
+
+#[async_trait]
+impl ResponsiveCheck for OptimisticResponsive {
+    async fn check_for_new_round(
+        &mut self,
+        highest_strong_links_round: Round,
+        _strong_links: Vec<NodeCertificate>,
+    ) {
+        let new_round = highest_strong_links_round + 1;
+        let _ = self.event_sender.send(new_round).await;
+    }
+
+    fn reset(&mut self) {}
+}
+
+enum State {
+    Initial,
+    Scheduled(JoinHandle<()>),
+    Sent,
+}
+
+/// More sophisticated strategy to move round forward given 2f+1 strong links
+/// Delay if backpressure is triggered. (TODO)
+/// Move as soon as 3f+1 is ready. (TODO: make it configurable)
+/// Move if minimal wait time is reached. (TODO: make it configurable)
+pub struct AdaptiveResponsive {
+    epoch_state: Arc<EpochState>,
+    start_time: Duration,
+    minimal_wait_time: Duration,
+    event_sender: tokio::sync::mpsc::Sender<Round>,
+    state: State,
+}
+
+impl AdaptiveResponsive {
+    pub fn new(
+        event_sender: tokio::sync::mpsc::Sender<Round>,
+        epoch_state: Arc<EpochState>,
+        minimal_wait_time: Duration,
+    ) -> Self {
+        Self {
+            epoch_state,
+            start_time: duration_since_epoch(),
+            minimal_wait_time,
+            event_sender,
+            state: State::Initial,
+        }
+    }
+}
+
+#[async_trait]
+impl ResponsiveCheck for AdaptiveResponsive {
+    async fn check_for_new_round(
+        &mut self,
+        highest_strong_links_round: Round,
+        strong_links: Vec<NodeCertificate>,
+    ) {
+        if matches!(self.state, State::Sent) {
+            return;
+        }
+        let new_round = highest_strong_links_round + 1;
+        observe_round(
+            self.start_time.as_micros() as u64,
+            RoundStage::StrongLinkReceived,
+        );
+        let voting_power = self
+            .epoch_state
+            .verifier
+            .sum_voting_power(strong_links.iter().map(|cert| cert.metadata().author()))
+            .expect("Unable to sum voting power from strong links");
+        // voting power == 3f+1 or pass minimal wait time
+        let duration_since_start = duration_since_epoch().saturating_sub(self.start_time);
+        if voting_power == self.epoch_state.verifier.total_voting_power()
+            || duration_since_start >= self.minimal_wait_time
+        {
+            let _ = self.event_sender.send(new_round).await;
+            if let State::Scheduled(handle) = std::mem::replace(&mut self.state, State::Sent) {
+                handle.abort();
```

Comment:
nit: abort first and then send? 

dag_driver handles duplicates, so its fine i think.